### PR TITLE
[WIP] Code coverage analysis

### LINKED ISSUE #81: Track test coverage with Coveralls
Free for open source, https://coveralls.io/pricing

> Coveralls is and will always be free for open source projects. You can always add a subscription to get coverage reports on private repos later.

Setup inspiration at https://github.com/sporkmonger/addressable/blob/295d7baec1281976434e9061b8065979b3f38c83/spec/spec_helper.rb

### PR DESCRIPTION
**Pull request by dentarg**
Locally with SimpleCov and on Travis CI with Coveralls.

Close #81.

### CODE REVIEW

**Comment by dentarg** [comment]
@twingly/dev WDYT?

File: spec/lib/twingly/url/hasher_spec.rb [hunk 1]
```diff
@@ -1,4 +1,4 @@
-require "spec_helper"
+SimpleCov.command_name(File.basename(__FILE__, ".rb"))
```
Comment (walro) [inline spec/lib/twingly/url/hasher_spec.rb]: I don't really understand what this is/why this is needed – do you have more info?

Comment (dentarg) [inline spec/lib/twingly/url/hasher_spec.rb]: Sure, I should have mentioned it. I has to do that we have split our tests over several different rake tasks and need to merge the results. I got it from https://github.com/colszowka/simplecov#merging-results. I tried doing `SimpleCov.command_name = "PID#{Process.pid}"` in `spec_helper.rb`, so we shouldn't have to add something to every spec file, but it didn't work well. It saved the results from every run, so it didn't discover regressions in coverage.

The command name can be seen in the footer in report:

> Generated by simplecov v0.12.0 and simplecov-html v0.10.0
using hasher_spec, null_url_spec, url_spec, utilities_spec

Comment (dentarg) [inline spec/lib/twingly/url/hasher_spec.rb]: > The command names can be seen in the footer in report:

Also in the spec run output: https://travis-ci.org/twingly/twingly-url/jobs/156254738

**Comment by walro** [comment]
I commented out all tests in `hasher_spec.rb` and I still got 100% coverage on `hasher.rb`.

**Comment by dentarg** [comment]
> I commented out all tests in hasher_spec.rb and I still got 100% coverage on hasher.rb.

Found a similar issue, https://github.com/colszowka/simplecov/issues/388, which is also using simplecov and coveralls, for a gem project, so I will take a look on what they've done

**Comment by dentarg** [comment]
Hmm, the comment at https://github.com/colszowka/simplecov/issues/314#issuecomment-48830291 was somewhat insightful:

>So, it's important understand how SimpleCov, and the stdlib Coverage lib it depends on works.

>It tracks when code is evaluated for the first time (i.e. when loaded or run )

>Thus, if you run SimpleCov.start, load a file, then run SimpleCov.result, you will see a certain amount of 'code coverage'. That is, running tests is a way to increase the coverage of a file above the coverage you get for just loading it.

>In your test helper you have `Dir[File.join(File.dirname(__FILE__), '../lib/**/*.rb')].sort.each { |f| require f }` which loads all your ruby folders under lib. Thus, you see files 'covered' that weren't tested.

I was thinking it might the combination of `lib/twingly/url/hasher.rb`, `lib/twingly/url/null_url.rb` and `lib/twingly/url/utilities.rb` all doing `require_relative "../url"`... and the merging of tests results. But I have to check some more.

Example: `bundle exec rspec spec/lib/twingly/url/hasher_spec.rb`, with only these lines in `hasher_spec.rb`:

```ruby
SimpleCov.command_name(File.basename(__FILE__, ".rb"))

require "twingly/url/hasher"
```

produces:

screen shot 2016-08-31 at 16 21 01

**Comment by dentarg** [comment]
I ran the specs one by one (no modifications) (`bundle exec rspec <spec file>`) and removed `tmp/coverage` before each run.

screen shot 2016-08-31 at 16 36 25

**Comment by dentarg** [comment]
@walro what do you think of just ignoring hasher and utilities for now?

```diff
diff --git a/spec/spec_helper.rb b/spec/spec_helper.rb
index f84a87a..a4b1734 100644
--- a/spec/spec_helper.rb
+++ b/spec/spec_helper.rb
@@ -7,6 +7,9 @@ SimpleCov.formatters = [
 SimpleCov.start do
   add_filter "spec/"
   add_filter "vendor/"
+
+  add_filter "hasher.rb"
+  add_filter "utilities.rb"
 end

 RSpec.configure do |config|
```

**Comment by walro** [comment]
> @walro what do you think of just ignoring hasher and utilities for now?

Hehe, nah. We could add new `.rb` files in the future that will exhibit the same weirdness so I think we'll need to find another route.

**Comment by dentarg** [comment]
Interesting? https://www.relishapp.com/rspec/rspec-core/v/3-5/docs/running-specs-multiple-times-with-different-runner-options-in-the-same-process

**Comment by walro** [comment]
I feel this is quite outdated now, I suspect things have moved forward since, close?

**Comment by dentarg** [comment]
> I feel this is quite outdated now, I suspect things have moved forward since, close?

Outdated how?

Sure, there has been a [decent amount of changes](https://github.com/twingly/twingly-url/compare/a2929f8...master) made since this was started, but our tests are still structured in the same way.

**Comment by walro** [comment]
Well, that's one part. The other part is that the tools have moved forward. Ruby has gained more things natively and so on. I feel that either we give this branch some love or just close it. It's approaching 18 months.

**Comment by dentarg** [comment]
Sure, it is probably best to start over.

I will note here what I think is an issue. Our setup is a bit special in that we run rspec multiple times when we run our spec suite (we started doing this to catch require errors):

```
$ bundle exec rake -T spec
rake spec:hasher_spec     # Run RSpec code examples
rake spec:null_url_spec   # Run RSpec code examples
rake spec:url_spec        # Run RSpec code examples
rake spec:utilities_spec  # Run RSpec code examples
```

https://github.com/twingly/twingly-url/blob/fedb25511bd1e5276310d507ef11645d78062e93/Rakefile#L19-L32

**Pull request closed without merging**